pkg/keycloak/keycloak.go

```diff
@@ -2,38 +2,196 @@ package keycloak
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal decoded parameters")
+		}
+
+		si := h.createServiceInstance(namespace, parameters, *sc)
+		serviceInstance, err := h.serviceCatalogClient.Servicecatalog().ServiceInstances(namespace).Create(&si)
+		if err != nil {
+			kcCopy.Status.Phase = v1alpha1.PhaseFailed
+			kcCopy.Status.Message = fmt.Sprintf("failed to create service instance: %v", err)
+
+			updateErr := sdk.Update(kcCopy)
+			if updateErr != nil {
+				return errors.Wrap(updateErr, fmt.Sprintf("failed to create service instance: %v, failed to update resource", err))
+			}
+
+			return errors.Wrap(err, "failed to create service instance")
+		}
+
+		kcCopy.Spec.InstanceID = serviceInstance.GetName()
+		kcCopy.Status.Phase = v1alpha1.PhaseProvisioning
+
+	case v1alpha1.PhaseProvisioning:
+		if kc.Spec.InstanceID == "" {
+			kcCopy.Status.Phase = v1alpha1.PhaseFailed
+			kcCopy.Status.Message = "instance ID is not defined"
+
+			err := sdk.Update(kcCopy)
+			if err != nil {
+				return errors.Wrap(err, "instance ID is not defined, failed to update resource")
+			}
+
+			return errors.New("instance ID is not defined")
+		} else {
+			si, err := h.serviceCatalogClient.Servicecatalog().ServiceInstances(namespace).Get(kc.Spec.InstanceID, metav1.GetOptions{})
+			if err != nil {
+				return errors.Wrap(err, "failed to get service instance")
+			}
+
+			if len(si.Status.Conditions) == 0 {
+				return nil
+			}
+
+			siCondition := si.Status.Conditions[0]
+			if siCondition.Type == "Ready" && siCondition.Status == "True" {
+				kcCopy.Status.Phase = v1alpha1.PhaseComplete
+				kcCopy.Status.Ready = true
+			}
+		}
+
+	case v1alpha1.PhaseDeprovisioning:
+		err := h.deleteKeycloak(kcCopy)
+		if err != nil {
+			kcCopy.Status.Phase = v1alpha1.PhaseDeprovisionFailed
+			kcCopy.Status.Message = fmt.Sprintf("failed to deprovision: %v", err)
+
+			updateErr := sdk.Update(kcCopy)
+			if updateErr != nil {
+				return errors.Wrap(updateErr, fmt.Sprintf("failed to deprovision instance: %v, failed to update resource", err))
+			}
+
+			return errors.Wrap(err, "failed to deprovision")
+		}
+
+		kcCopy.Status.Phase = v1alpha1.PhaseDeprovisioned
+
+	case v1alpha1.PhaseDeprovisioned:
+		kcCopy.Finalizers = []string{}
+	}
+
+	// Only update the Keycloak custom resource if there was a change
+	if !reflect.DeepEqual(kc, kcCopy) {
+		if err := sdk.Update(kcCopy); err != nil {
+			return errors.Wrap(err, "failed to update the keycloak resource")
+		}
+	}
+
 	// set up authenticated client
 	authenticatedClient, err := h.kcClientFactory.AuthenticatedClient(*kcCopy)
 	if err != nil {
 		return errors.Wrap(err, "failed to get authenticated client for keycloak")
 	}
+	// hand of each realm to reconcile realm may want to make async to avoid blocking
+	for _, r := range kcCopy.Spec.Realms {
+		if err := h.reconcileRealm(ctx, r, authenticatedClient); err != nil {
+			return errors.Wrap(err, "failed to reconcile realm "+r.Name)
+		}
+	}
 
 	if event.Deleted {
```

Comment:
Is there anything that will go here in the future? Otherwise, can we delete this?